### PR TITLE
Add `i1` type propagation for named Linalg ops.

### LINKED ISSUE #12446: Type Propagation fails with i1, i1 -> i1 `linalg.matmul` matmul
Executing `iree-opt --iree-codegen-type-propagation` with the block below results in illegal IR. This is likely due to type propagation within the `linalg.matmul` block.

```mlir
func.func @main_dispatch_4_matmul_6x3x8(%arg0 : tensor<6x8xi8>, %arg1 : tensor<8x3xi8>) -> tensor<6x3xi8> {
  %c0 = arith.constant 0 : index
  %c128 = arith.constant 128 : index
  %c64 = arith.constant 64 : index
  %false = arith.constant false
  %4 = arith.trunci %arg0 : tensor<6x8xi8> to tensor<6x8xi1>
  %6 = arith.trunci %arg1 : tensor<8x3xi8> to tensor<8x3xi1>
  %7 = tensor.empty() : tensor<6x3xi1>
  %8 = linalg.fill ins(%false : i1) outs(%7 : tensor<6x3xi1>) -> tensor<6x3xi1>
  %9 = linalg.matmul ins(%4, %6 : tensor<6x8xi1>, tensor<8x3xi1>) outs(%8 : tensor<6x3xi1>) -> tensor<6x3xi1>
  %10 = arith.extui %9 : tensor<6x3xi1> to tensor<6x3xi8>
  return %10 : tensor<6x3xi8>
}
```

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
Current propagation didnt handle cases where named ops have `i1` type. Since the named op body is different based on the element type, the region of the operation cannot be cloned. Instead, a new named op of the same type is created.

Fixes #12446

### CODE REVIEW

**Comment by rsuderman** [comment]
Curious question - what is the effect on the region of the `linalg.matmul`? With the `i1` `linalg.matmul` implementation we perform as `arith.and` and `arith.or` while the `i8` implementation does `arith.mul` and `arith.add`. Does the change actually change the contents of the internal loop? The difference feels minor however it would result in potentially overflowing. Not 100% certain how likely this is.

File: compiler/src/iree/compiler/Codegen/Common/TypePropagationPass.cpp [hunk 1]
```diff
@@ -142,6 +142,29 @@ struct ConstantOpTypeConversion
...
+  using TypePropagationPattern<OpTy>::TypePropagationPattern;
+
+  LogicalResult matchAndRewrite(
+      OpTy namedOp, typename OpTy::Adaptor adaptor,
+      ConversionPatternRewriter &rewriter) const final {
+    SmallVector<Type> resultTypes;
+    resultTypes.reserve(namedOp->getNumResults());
+    for (auto resultType : namedOp->getResultTypes()) {
+      Type legalizedType = this->getTypeConverter()->convertType(resultType);
+      resultTypes.push_back(resultType);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/TypePropagationPass.cpp]: legalizedType is unused. Should we append it to resultTypes?

**Comment by hanhanW** [comment]
> Curious question - what is the effect on the region of the `linalg.matmul`? With the `i1` `linalg.matmul` implementation we perform as `arith.and` and `arith.or` while the `i8` implementation does `arith.mul` and `arith.add`. Does the change actually change the contents of the internal loop? The difference feels minor however it would result in potentially overflowing. Not 100% certain how likely this is.

I think it is changing the content of the body. The rewriter creates a new matmul op. In this context, a Matmul::regionBuilder will be called. It would determine the types and create corresponding ops. E.g.,

```
Matmul::regionBuilder(...) {
  ...
  Value value1 = helper.buildTypeFn(castVal, block.getArgument(2).getType(), block.getArgument(0));
  Value value2 = helper.buildTypeFn(castVal, block.getArgument(2).getType(), block.getArgument(1));
  Value value3 = helper.buildBinaryFn(BinaryFn::mul, value1, value2);
  Value value4 = helper.buildBinaryFn(BinaryFn::add, block.getArgument(2), value3);
}
```

And one example in buildBinaryFn:

```
    case BinaryFn::add:
      if (allComplex)
        return builder.create<complex::AddOp>(arg0.getLoc(), arg0, arg1);
      if (allFloatingPoint)
        return builder.create<arith::AddFOp>(arg0.getLoc(), arg0, arg1);
      if (allBool)
        return builder.create<arith::OrIOp>(arg0.getLoc(), arg0, arg1);
      return builder.create<arith::AddIOp>(arg0.getLoc(), arg0, arg1);
```

File: compiler/src/iree/compiler/Codegen/Common/TypePropagationPass.cpp (hunk 1 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Common/TypePropagationPass.cpp]: Uggh. Obviously I need to use `legalizedType`. Why this worked is beyond me right now.

**Comment by MaheshRavishankar** [comment]
> Curious question - what is the effect on the region of the `linalg.matmul`? With the `i1` `linalg.matmul` implementation we perform as `arith.and` and `arith.or` while the `i8` implementation does `arith.mul` and `arith.add`. Does the change actually change the contents of the internal loop? The difference feels minor however it would result in potentially overflowing. Not 100% certain how likely this is.

What Hanhan said, but why would it overflow? The original code was i1. Now instead of an `and`/`or` you do a `mul`/`add`. The value is still just 0 or 1, it is just using `i8` isntead of `i1`. `i1` load store creates issues. So we just head this off very early on...

**Review by hanhanW — APPROVED** [review]
assuming that the comment will be addressed

**Pull request merged**